### PR TITLE
Add mask identifier for multiplexed src_mask/src_key_padding_mask in BT

### PR DESCRIPTION
**Pull request by nmacchioni**
Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/81947

Transformer fastpath multiplexes two arguments, src_mask [seq_len x seq_len] and src_key_padding_mask [batch_size x seq_len], and later deduces the type based on mask shape.

In the event that batch_size == seq_len, any src_mask is wrongly interpreted as a src_key padding_mask. This is fixed by requiring a mask_type identifier be supplied whenever batch_size == seq_len.

Additionally, added support for src_mask in masked_softmax CPU path.

Test Plan: existing unit tests + new unit tests (batch_size == seq_len)

Differential Revision: D37932240

### CODE REVIEW

**Review by zrphercule — COMMENTED** [review]
Overall LGTM, comments in below

File: aten/src/ATen/native/cuda/SoftMax.cu [hunk 1]
```diff
@@ -957,18 +957,19 @@ TORCH_IMPL_FUNC(softmax_backward_cuda_out)
...
-Tensor masked_softmax_cuda(const Tensor& input_, const Tensor& mask_, const c10::optional<int64_t> dim_) {
+Tensor masked_softmax_cuda(const Tensor& input_, const Tensor& mask_, const c10::optional<int64_t> dim_, const c10::optional<int64_t> mask_type_) {
   Tensor output = at::empty_like(input_, input_.options());
   TORCH_CHECK(mask_.scalar_type() == ScalarType::Bool, "Mask should be a boolean tensor");
+  TORCH_CHECK(mask_type_, "expected `mask_type` when `mask` supplied");
 
   // If input is [B, H, T, T] and mask is [B, T]
   // we have special fast kernel
-  bool is_BxT_mask = (input_.dim() == 4 && mask_.dim() == 2 && input_.size(0) == mask_.size(0) && input_.size(2) == mask_.size(1) && input_.size(3) == mask_.size(1));
+  bool is_BxT_mask = (mask_type_ == 2) && (input_.dim() == 4 && mask_.dim() == 2 && input_.size(0) == mask_.size(0) && input_.size(2) == mask_.size(1) && input_.size(3) == mask_.size(1));
```
Comment (zrphercule) [inline aten/src/ATen/native/cuda/SoftMax.cu:967]: Can we use Enum instead of hardcoded number here? Or at least we need some comments indicate what mask_type==1 or ==2 means

File: aten/src/ATen/native/SoftMax.cpp [hunk 2]
```diff
@@ -131,7 +131,10 @@ void host_softmax(
     Tensor output,
     const Tensor& input,
     const int64_t dim,
-    bool* mask = nullptr) {
+    bool* mask = nullptr,
+    const c10::optional<int64_t> mask_type = NULL) {
+  TORCH_CHECK(!mask || mask_type,
```
Comment (zrphercule) [inline aten/src/ATen/native/SoftMax.cpp:136]: Please make sure adding mask_type for all existing code

File: aten/src/ATen/native/cuda/SoftMax.cu (hunk 1 above)
Comment (nmacchioni) [reply aten/src/ATen/native/cuda/SoftMax.cu:967]: I was originally looking to use an Enum but couldn't figure out where the best location for it was? If you had a suggestion that would be appreciated and then I could go back and change the types.

Comment (zrphercule) [reply aten/src/ATen/native/cuda/SoftMax.cu:967]: > 
Since this is being passed as param of function, I guess you can keep it as number. But please put some comments at the beginning of the function, to clarify each number's meaning.

**Comment by nmacchioni** [comment]
@zrphercule Could you please approve the above workflows

**Review by zrphercule — APPROVED** [review]
Thanks!

File: aten/src/ATen/native/SoftMax.cpp [hunk 3]
```diff
@@ -131,7 +131,18 @@ void host_softmax(
...
     const Tensor& input,
     const int64_t dim,
-    bool* mask = nullptr) {
+    bool* mask = nullptr,
+    const c10::optional<int64_t> mask_type_ = NULL) {
+
+  if (MaskedSoftMax) {
+    TORCH_CHECK(mask_type_.has_value(), "Mask Type should be defined");
+    int64_t mask_type = mask_type_.value();
+    TORCH_CHECK((mask_type == 0) || (mask_type == 1), "Mask Type should be 0 (src_mask) or 1 (src_key_padding_mask)");
```
Comment (zrphercule) [inline aten/src/ATen/native/SoftMax.cpp:140]: Maybe re-org this part of logic a bit? since mask_type == 0 does not work on CPU

File: test/test_nn.py [hunk 4]
```diff
@@ -20862,6 +20867,17 @@ def test_multihead_attn_in_proj_weight_none(self, device, dtype):
         key = torch.rand(4, 4, 2, dtype=dtype, device=device)
         mha(query, key, key)
 
+    @onlyCPU
```
Comment (zrphercule) [inline test/test_nn.py:20870]: Why only CPU?

**Review by zrphercule — COMMENTED** [review]
I feel this diff itself overall LGTM, but will it break internal CI? Since you changed some API

**Review by zrphercule — APPROVED** [review]
OK CI looks fine, thanks

File: aten/src/ATen/native/SoftMax.cpp (hunk 3 above)
Comment (nmacchioni) [reply aten/src/ATen/native/SoftMax.cpp:140]: Since adding support for mask_type == 0 is in progress I chose to have that check separate as it will be removed soon

File: test/test_nn.py (hunk 4 above)
Comment (nmacchioni) [reply test/test_nn.py:20870]: Test is specifically targeted at addressing issue #81129 which only occurs on CPU
